### PR TITLE
Change profile matching to match what the web interface says.

### PR DESCRIPTION
**Pull request by BaronGreenback**
https://github.com/jellyfin/jellyfin/issues/5854

Web interfaces says""Enter at least one identification criteria".

Currently profiles are being matched on **all** criteria.

This PR - changes the profile matching to what the web says and includes a test.

### CODE REVIEW

File: Emby.Dlna/DlnaManager.cs [hunk 1]
```diff
@@ -106,19 +106,28 @@ public DeviceProfile GetProfile(DeviceIdentification deviceInfo)
...
-            if (profile != null)
+            try
             {
-                _logger.LogDebug("Found matching device profile: {0}", profile.Name);
+                var profile = GetProfiles()
+                    .FirstOrDefault(i => i.Identification != null && IsMatch(deviceInfo, i.Identification));
+
+                if (profile != null)
+                {
+                    _logger.LogDebug("Found matching device profile: {0}", profile.Name);
```
Comment (cvium) [inline Emby.Dlna/DlnaManager.cs:116]:
```suggestion
                    _logger.LogDebug("Found matching device profile: {ProfileName}", profile.Name);
```

File: Emby.Dlna/DlnaManager.cs [hunk 2]
```diff
@@ -138,85 +147,49 @@ private void LogUnmatchedProfile(DeviceIdentification profile)
...
             {
-                return input.Contains(pattern, StringComparison.OrdinalIgnoreCase) || Regex.IsMatch(input, pattern, RegexOptions.IgnoreCase | RegexOptions.CultureInvariant);
+                return input.Equals(pattern, StringComparison.OrdinalIgnoreCase)
+                    || Regex.IsMatch(input, pattern, RegexOptions.IgnoreCase | RegexOptions.CultureInvariant);
             }
             catch (ArgumentException ex)
             {
-                _logger.LogError(ex, "Error evaluating regex pattern {Pattern}", pattern);
-                return false;
+                throw new ArgumentException("Error evaluating regex pattern " + pattern, ex);
```
Comment (cvium) [inline Emby.Dlna/DlnaManager.cs:192]: Is throwing necessary?

Comment (cvium) [reply Emby.Dlna/DlnaManager.cs:192]: > To get around the issue of having to write multiple mock classes for the testing unit, I made this method static.

Why would you need to write mock classes? The DlnaManager constructor doesn't do anything, so you could just mock all its dependencies with Moq in your test and then `new` the manager.

Comment (BaronGreenback) [reply Emby.Dlna/DlnaManager.cs:192]: just me being a lazy ass - have changed the code.

File: Emby.Dlna/DlnaManager.cs [hunk 3]
```diff
@@ -138,80 +138,45 @@ private void LogUnmatchedProfile(DeviceIdentification profile)
...
-
-            if (!string.IsNullOrEmpty(profileInfo.ModelNumber))
-            {
-                if (deviceInfo.ModelNumber == null || !IsRegexOrSubstringMatch(deviceInfo.ModelNumber, profileInfo.ModelNumber))
-                {
-                    return false;
-                }
-            }
+            return IsRegexOrSubstringMatch(deviceInfo.FriendlyName, profileInfo.FriendlyName)
+               && IsRegexOrSubstringMatch(deviceInfo.Manufacturer, profileInfo.Manufacturer)
```
Comment (cvium) [inline Emby.Dlna/DlnaManager.cs:153]: Indentation seems a little inconsistent. It's three spaces here, but in `IsRegexOrSubstringMatch` it's four?

File: tests/Jellyfin.Dlna.Tests/ProfileTester.cs [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+                    FriendlyName = "My Device",
+                    Manufacturer = "LG Electronics",
+                    ManufacturerUrl = "http://www.lge.com",
+                    ModelDescription = "LG WebOSTV DMRplus",
+                    ModelName = "LG TV",
+                    ModelNumber = "1.0",
+                }
+            };
+
+            Assert.True(GetManager().IsMatch(device.ToDeviceIdentification(), profile.Identification));
```
Comment (cvium) [inline tests/Jellyfin.Dlna.Tests/ProfileTester.cs:60]: Typically we use the three A's:

1. Arrange
2. Act
3. Assert

Can you move the assertions down to the bottom?

File: tests/Jellyfin.Dlna.Tests/ProfileTester.cs [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+using MediaBrowser.Model.Dlna;
+using MediaBrowser.Model.IO;
+using MediaBrowser.Model.Serialization;
+using Microsoft.Extensions.Logging;
+using Moq;
+using Xunit;
+
+namespace Jellyfin.Dlna.Tests
+{
+    public class ProfileTester
```
Comment (cvium) [inline tests/Jellyfin.Dlna.Tests/ProfileTester.cs:14]: Alternatively if it becomes too unwieldy in the future, we can move the tests into a DlnaManager folder and rename the tests.
```suggestion
    public class DlnaManagerTests
```

File: tests/Jellyfin.Dlna.Tests/ProfileTester.cs [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+            var fileSystem = new Mock<IFileSystem>();
+            var appPaths = new Mock<IApplicationPaths>();
+            var loggerFactory = new Mock<ILoggerFactory>();
+            var appHost = new Mock<IServerApplicationHost>();
+
+            return new DlnaManager(xmlSerializer.Object, fileSystem.Object, appPaths.Object, loggerFactory.Object, appHost.Object);
+        }
+
+        [Fact]
+        public void Test_Profile_Matches()
```
Comment (cvium) [inline tests/Jellyfin.Dlna.Tests/ProfileTester.cs:28]: No need to put `Test` in the method name, we know it's a test. We also try to adhere to the `MethodName_Scenario_ExpectedResult` naming scheme.
```suggestion
        public void IsMatch_GivenMatchingName_ReturnsTrue()
```

File: tests/Jellyfin.Dlna.Tests/ProfileTester.cs [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+                {
+                    FriendlyName = "My Device",
+                }
+            };
+
+            Assert.True(GetManager().IsMatch(device.ToDeviceIdentification(), profile2.Identification));
+        }
+
+        [Fact]
+        public void Test_Profile_NoMatch()
```
Comment (cvium) [inline tests/Jellyfin.Dlna.Tests/ProfileTester.cs:76]:
```suggestion
        public void IsMatch_GivenNamesAndManufacturersDoNotMatch_ReturnsFalse()
```

**Review by cvium — CHANGES_REQUESTED** [review]

**Comment by cvium** [comment]
Nice to have: Regex test.

File: tests/Jellyfin.Dlna.Tests/DlnaManagerTests.cs [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+                    FriendlyName = "My Device",
+                    Manufacturer = "LG Electronics",
+                    ManufacturerUrl = "http://www.lge.com",
+                    ModelDescription = "LG WebOSTV DMRplus",
+                    ModelName = "LG TV",
+                    ModelNumber = "1.0",
+                }
+            };
+
+            Assert.True(GetManager().IsMatch(device.ToDeviceIdentification(), profile.Identification));
```
Comment (cvium) [inline tests/Jellyfin.Dlna.Tests/DlnaManagerTests.cs:60]: You forgot to move it

**Review by cvium — APPROVED** [review]

**Review by crobibero — APPROVED** [review]

**Pull request merged**